### PR TITLE
CMake: enable dynamic linking for Windows

### LINKED ISSUE #2460: Windows dynamic linking
Investigate dynamic linking under windows:

https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/
https://bitbucket.org/vlasovmaksim/cmake_windows_export_all_symbols/overview
https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html

### PR DESCRIPTION
**Pull request by tamiko**
Let's see what happens.
Reference: https://www.kitware.com//create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/

Closes #2460

### CODE REVIEW

**Comment by tamiko** [comment]
:crossed_fingers:

**Comment by tjhei** [comment]
This is indeed epic if it works! :+1:

**Comment by drwells** [comment]
> LINK : fatal error LNK1189: library limit of 65535 objects exceeded [D:\a\dealii\dealii\build\source\dealii_debug.vcxproj]

:(

**Comment by tjhei** [comment]
This error looks familiar. I think I tried that many years ago.

**Comment by tamiko** [comment]
The problem is the simple fact that a Windows `dll` has a 16bit descriptor for symbol objects which makes for a total of 2^16 symbols. My deal.II library (with almost everything enabled) currently has around 270k symbols defined.

There is no way around that ABI limitation short of
 - either manually controlling what symbols have to be exported (which is not going to happen),
 - split the library itself into multiple parts.
The latter would require that our current object targets would form an acyclical graph. Which they do not.

**Comment by tjhei** [comment]
> * The latter would require that our current object targets would form an acyclical graph. Which they do not.

While this sounds neat, I don't think it is realistic to untangle them. Maybe we could reshuffle functionality to end up with 2-3 libraries, but that would not be enough either.

**Comment by tamiko** [comment]
@tjhei Agreed. Unfortunately :sob:

File: cmake/checks/check_02_system_features.cmake [hunk 1]
```diff
@@ -100,21 +99,10 @@ if(CMAKE_SYSTEM_NAME MATCHES "Windows")
...
-      )
-    set(BUILD_SHARED_LIBS OFF CACHE BOOL "" FORCE)
-
-    # And disable compilation of examples:
-    set(DEAL_II_COMPILE_EXAMPLES OFF CACHE BOOL "" FORCE)
+  # Let's use the CMake infrastructure to automatically export symbols on
+  # Windows targets:
+  #
+  if(BUILD_SHARED_LIBS)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
```
Comment (bangerth) [inline cmake/checks/check_02_system_features.cmake:106]: @marcfehling Relating to #13596, I think this block (or a part of it) might still make for a good independent patch. I know that the rest of the current PR does not work, but this part is at least conceptually correct even if it does not have any effect right now.
